doc: note the costs of fetching all pull requests

vasild opened the pull request:
Also mention that it is possible to fetch just one pull request.

kristapsk commented on doc/productivity.md:
````diff
@@ -176,7 +176,9 @@ When looking at other's pull requests, it may make sense to add the following se
 ```
 
-This will add an `upstream-pull` remote to your git repository, which can be fetched using `git fetch --all` or `git fetch upstream-pull`. Afterwards, you can use `upstream-pull/NUMBER/head` in arguments to `git show`, `git checkout` and anywhere a commit id would be acceptable to see the changes from pull request NUMBER.
+This will add an `upstream-pull` remote to your git repository, which can be fetched using `git fetch --all` or `git fetch upstream-pull`. It will download ~800 MB (all PRs, including merged and closed ones) and will expand the disk usage from ~180 MB to ~1400 MB. Afterwards, you can use `upstream-pull/NUMBER/head` in arguments to `git show`, `git checkout` and anywhere a commit id would be accep ...
````
Maybe it's better to not specify exact numbers and use something like "hundreds of megabytes" instead, as it will change in future?

vasild replied:
Done. Shortened to just "It will download and store on disk about 1 GB of data" - enough to avoid the surprise.

MarcoFalke commented:
I tend to fetch only a single pull by just copying the URL in blue on the top right (e.g. for this pull https://github.com/vasild/bitcoin/tree/note_fetch_all_cost)

Then, paste it into a terminal and replace the `tree/` with a space:

```
$ git fetch https://github.com/vasild/bitcoin/ note_fetch_all_cost

remote: Enumerating objects: 14, done.
remote: Counting objects: 100% (14/14), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 17 (delta 11), reused 13 (delta 11), pack-reused 3
Unpacking objects: 100% (17/17), 10.37 KiB | 212.00 KiB/s, done.
From https://github.com/vasild/bitcoin
 * branch                  note_fetch_all_cost -> FETCH_HEAD

vasild commented:
Actually, isn't it an overkill to download all pull requests - opened, closed and merged (~1 GB)? Usually one wants to review or test one PR.

@sipa you added this in https://github.com/bitcoin/bitcoin/commit/31444491f2#diff-e4de6d71249c78e229a61d6a03ac255aR422 almost 4 years ago when the volume of this was smaller. What about dropping this advice from the doc and replace it with instructions on fetching a single PR, as per [the comment above](https://github.com/bitcoin/bitcoin/pull/18382#issuecomment-601191758) and/or [github help](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally#modifying-an-inactive-pull-request-locally)?

MarcoFalke commented:
Some people might have a really large disk and review a lot of pull requests or want to archive everything that happens for other reasons, so it might be good to mention both: the "archival solution" and the quick fetch solution.

fjahr commented on doc/productivity.md:
```diff
@@ -176,7 +176,9 @@ When looking at other's pull requests, it may make sense to add the following se
-This will add an `upstream-pull` remote to your git repository, which can be fetched using `git fetch --all` or `git fetch upstream-pull`. Afterwards, you can use `upstream-pull/NUMBER/head` in arguments to `git show`, `git checkout` and anywhere a commit id would be acceptable to see the changes from pull request NUMBER.
+This will add an `upstream-pull` remote to your git repository, which can be fetched using `git fetch --all` or `git fetch upstream-pull`. It will download ~800 MB (all PRs, including merged and closed ones) and will expand the disk usage from ~180 MB to ~1400 MB. Afterwards, you can use `upstream-pull/NUMBER/head` in arguments to `git show`, `git checkout` and anywhere a commit id would be accep ...
+
+It is also possible to [fetch only a given pull request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally#modifying-an-inactive-pull-request-locally).
```
I think referencing that link here is fragile. I would rather have the fetch example cli command.

vasild replied:
Agree, removed the link.

fjahr reviewed (commented):
Concept ACK

I think this is good so people don't get surprised by the large amount of data. Agree with @kristapsk that there should not be any specific numbers though.

vasild commented:
Updated: https://github.com/bitcoin/bitcoin/blob/5ebd8f6b84eeff3bc3da219aa7ce978d861db65d/doc/productivity.md#reference-prs-easily-with-refspecs

theStack reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/18382/commits/5ebd8f6b84eeff3bc3da219aa7ce978d861db65d
Thanks for adding those valuable informations/instructions :+1:

promag commented:
FWIW I also tend to add a remote for each recurring contributor fork which also allows me to sneak peak what they are doing.

laanwj commented:
This renders as:

![screenshot](https://user-images.githubusercontent.com/126646/77678671-cab4a980-6f91-11ea-9955-397ba7f65d80.png)

I don't think the "For example:"s should be in the code box.

vasild commented:
Right, updated. It [looks better now](https://github.com/bitcoin/bitcoin/blob/abf7ff177/doc/productivity.md#reference-prs-easily-with-refspecs).

sipa commented:
Doing some numbers myself:

Cloning bitcoin/bitcoin downloads 138.8 MiB, resulting in a size on disk of .git of 146.8 MB.
Starting with a fresh clone:
* Fetching refs/pull/* downloads an additional 801.37 MiB, resulting in a size on disk of .git of 1405.4 MB.
* Fetching refs/pull/*/head instead downloads an additional 797.09 MiB, resulting in a size on disk of .git of 1233.0 MB.

I've generally never had a use for the refs/pull/*/merge branches (I prefer doing a merge with master myself if needed anyway, and that's rare), maybe it's reasonable to only suggest fetching the heads?

vasild commented:
Those numbers more or less coincide with my observations.
I agree with `refs/pull/*` vs `refs/pull/*/head`.
Updated, [render preview](https://github.com/bitcoin/bitcoin/blob/61dd7e2ca/doc/productivity.md#reference-prs-easily-with-refspecs).

laanwj commented:
> Right, updated. It looks better now.

Thanks, yes, looks much better now.

jonatack commented on doc/productivity.md:
````diff
@@ -172,11 +172,35 @@ When looking at other's pull requests, it may make sense to add the following se
+for example:
+
+```
+git fetch https://github.com/glowang/bitcoin/ update_blocksonly_docs:docs_pr
````
Have you checked with this person about using their identity here?

sipa replied:
Ping @glowang.

glowang replied:
thank you for asking - this is fine with me!

jonatack reviewed (commented):
As there are many ways to do this, and a wide range of user computing configurations and needs, it never really made sense to me why these precise instructions (and now examples) would be provided. (I use a somewhat different gitconfig for this and know other contributors who do as well). Just a counterpoint.

RandyMcMillan commented:
It may be useful to add a note about pulling from the most recent passing commit hash on Travis-ci.

laanwj commented:
> It may be useful to add a note about pulling from the most recent passing commit hash on Travis-ci.

I don't think github allows fetching by commit hash. You can only fetch labels or branches. It would be awesomely convenient sometimes to retrieve an old version of a PR that has been force-pushed over, to compare, but I've never managed to do this.

promag commented:
> I don't think github allows fetching by commit hash. You can only fetch labels or branches.

> FWIW I also tend to add a remote for each recurring contributor fork which also allows me to sneak peek what they are doing.

@laanwj I do `git remote update laawnj && git reset --hard laawnj/<sha>`

theStack reviewed (approved):
re-ACK https://github.com/bitcoin/bitcoin/pull/18382/commits/61dd7e2ca7dcfe519023bc4882777d348d1f70fd
Just checked that since my last ACK:
* formatting has improved (the `for example:` string is now outside of the monospaced code block)
* example `.gitconfig` section has been changed to only retreiving the `head`s

theStack commented:
> As there are many ways to do this, and a wide range of user computing configurations and needs, it never really made sense to me why these precise instructions (and now examples) would be provided. (I use a somewhat different gitconfig for this and know other contributors who do as well). Just a counterpoint.

Personally speaking, the instructions on how to fetch all pull requests were very helpful when I started contributing/reviewing. It would be highly absurd to encourage new contributors to review and test code but then not including any instructions how to do so in terms of git(hub) workflow IMHO. I cannot follow the argumentation "there are many ways to do it, so better not even describe a single one of them".

laanwj commented:
> @laanwj I do git remote update laawnj && git reset --hard laawnj/<sha>

You don't need the `laanwj/` before the hash. But yes, that will work, but only for commits that are currently (or have been at some point when you've fetched) connected to a tag or branch. If you've missed the point they were part of a branch, you cannot fetch them anymore. You can however view those commits through the web interface.

> Personally speaking, the instructions on how to fetch all pull requests were very helpful when I started contributing/reviewing.

Personally I think it's useful to have some tips for things people frequently want to do, but otherwise refer to existing documentation. I also understand @jonatack's point that we don't want to end up maintaining an alternative git documentation here.

jonatack commented:
> It would be highly absurd to encourage new contributors to review and test code but then not including any instructions how to do so in terms of git(hub) workflow IMHO. I cannot follow the argumentation "there are many ways to do it, so better not even describe a single one of them".

I and others do describe these things more or less in detail in various [articles](https://jonatack.github.io/articles) to help onboard new people, a process I've been going through as well. Also, there are resources like https://bitcoincore.reviews to help onboard contributors -- and which I continue to learn from -- which everyone can [contribute to](https://github.com/bitcoin-core-review-club/website/graphs/contributors) and participate in.

theStack commented:
> > It would be highly absurd to encourage new contributors to review and test code but then not including any instructions how to do so in terms of git(hub) workflow IMHO. I cannot follow the argumentation "there are many ways to do it, so better not even describe a single one of them".
> 
> I and [others](https://github.com/fanquake/core-review) do describe these things more or less [in detail](https://jonatack.github.io/articles/how-to-review-pull-requests-in-bitcoin-core#pull-down-the-code-locally) in various [articles](https://jonatack.github.io/articles) to help onboard new people, a process I've been going through as well. Also, there are resources like https://bitcoincore.reviews to help onboard contributors -- and which I continue to learn from -- which everyone can [contribute to](https://github.com/bitcoin-core-review-club/website/graphs/contributors) and participate in.

Fair enough, there is indeed very good and deep introductory material out there (for me it was [this article by Jimmy Song](https://bitcointechtalk.com/a-gentle-introduction-to-bitcoin-core-development-fdc95eaee6b8) that got me started) and the PR Review Club is an absolutely awesome idea. I'd still without hesitation recommend to every newcomer to carefully read the included `doc`s as early as possible (at the very least `developer-notes.md` and `productivity.md`), as common denominator of contributor-agreed best practices and tips, that are less likely to get outdated than articles maintained by single individuals. To sum up my view: as long as we _have_ documents like `productivity.md` included in the repository, they should also include basics about how to fetch and checkout pull requests.

jonatack commented on doc/productivity.md:
````diff
@@ -172,11 +172,35 @@ When looking at other's pull requests, it may make sense to add the following se
 ```
 
-This will add an `upstream-pull` remote to your git repository, which can be fetched using `git fetch --all` or `git fetch upstream-pull`. Afterwards, you can use `upstream-pull/NUMBER/head` in arguments to `git show`, `git checkout` and anywhere a commit id would be acceptable to see the changes from pull request NUMBER.
+This will add an `upstream-pull` remote to your git repository, which can be fetched using `git fetch --all` or `git fetch upstream-pull`. It will download and store on disk about 1 GB of data (all PRs, including merged and closed ones). Afterwards, you can use `upstream-pull/NUMBER/head` in arguments to `git show`, `git checkout` and anywhere a commit id would be acceptable to see the changes fr ...
````
With respect to "about 1 GB of data", I'd suggest not stating an amount needs to be updated/maintained.

vasild replied:
Reworded to "quite a lot of data", its good enough.

vasild committed:
doc: note the costs of fetching all pull requests

Also, update the example to skip downloading the merge commits when
downloading all PRs.

vasild commented:
This PR caused more discussions than I anticipated plus some controversy.

The initial reason I opened it is that I expected the `upstream-pull` remote to download only the currently opened PRs and the download to be less than 100-ish MB, so I wanted to "warn" others that it is more expensive than that, given that git does not display the total download size upfront.

Updated the PR with just this basic info in one sentence (plus download only heads).

[Render preview](https://github.com/bitcoin/bitcoin/blob/d695eb4c2/doc/productivity.md#reference-prs-easily-with-refspecs)

MarcoFalke commented:
ACK d695eb4c2164fc8777dc014e1a30c014cf04982a

fanquake reviewed (approved):
ACK d695eb4c2164fc8777dc014e1a30c014cf04982a
